compiler/noirc_evaluator/src/ssa/interpreter/mod.rs

```diff
@@ -124,19 +128,20 @@ impl<'ssa, W: Write> Interpreter<'ssa, W> {
         output: W,
     ) -> Self {
         let call_stack = vec![CallContext::global_context()];
-        Self { functions, call_stack, side_effects_enabled: true, options, output, step_counter: 0 }
+        Self { functions, call_stack, options, output, step_counter: 0 }
     }
 
     pub(crate) fn functions(&self) -> &BTreeMap<FunctionId, Function> {
         self.functions
     }
 
-    /// Resets the step counter to 0.
+    /// Resets interpreter state for reuse between top-level calls.
     ///
-    /// This can be used when the interpreter is reused between calls,
-    /// to reset the budget before interpreting the next entry point.
-    pub(crate) fn reset_step_counter(&mut self) {
+    /// This resets the step counter and side_effects_enabled state,
+    /// preparing the interpreter to interpret a new entry point.
+    pub(crate) fn reset_state(&mut self) {
         self.step_counter = 0;
+        self.call_context_mut().side_effects_enabled = true;
```

Comment:
If it's based on the call context, I don't think it needs to be restored, does it?

OTOH, I'm not sure if the interpreter returns an `Err` such as a constraint failure, does the call context stack get cleaned up? I don't think it does. So instead of modifying it, the `call_stack` should be truncated back to contain only the first (global) context.

Reply:
brillig calls are not impacted by side_effects
Also, it will push a new call context for each function.

Reply:
True, but the `call_stack` contains variable bindings as well. If all is well I expect a variable lookup would not bubble up beyond the entry of the current call, however we are 1) leaking memory to some extent by not freeing resources and 2) making it harder to reach the global context, because we have to skip over all the failed call stacks until we find their variable ID in the first frame (although it might realise it's a global and go there directly, I'm not sure).

Reply:
Sorry, there is no transitive lookups in the `scope`, so that's not a problem. Then we are just growing it with each failure, and the only observable side effect would probably be when the current top level call ends, it would print "back in function ..." with whatever the previously failed function was.